lib/task_helpers/development/replication.rb

```diff
@@ -0,0 +1,126 @@
+        end
+
+        def restore
+          if File.exist?(backup_guid)
+            FileUtils.rm_f(guid_file)
+            FileUtils.mv(backup_guid, guid_file)
+          end
+        end
+
+        def setup
+          REMOTES.each {|r| setup_remote(r) }
+          setup_global(GLOBAL)
+
+          # TODO: We have the technology to watch for this and report when it's all good or bad
+          puts "Local replication is setup... try checking for users in all regions: psql -U root development_replication_99 -c \"select id from users;\""
+        end
+
+        def teardown
+          REMOTES.each do |r|
+            teardown_global_subscription_for_region(r)
+            teardown_remote_publication(r)
+          end
+        end
+
+        def setup_remote(region)
+          setup_one_region(region)
+          setup_remote_region(region)
+        end
+
+        def setup_global(region)
+          setup_one_region(region)
+          setup_global_region(region)
+        end
+
+        def database(region)
+          "#{DB_PREFIX}_#{region}"
+        end
+
+        # Example: DATABASE_URL='postgres://root:smartvm@localhost:5432/development_replication_99'
+        def database_url(region)
+          "postgres://#{PG_USER}:#{PG_PASS}@#{PG_HOST}:#{PG_PORT}/#{database(region)}"
+        end
+
+        def guid_file
+          GUID_FILE
+        end
+
+        def backup_guid
+          BACKUP_GUID
+        end
+
+        def setup_global_region_script
+          subs = []
+          REMOTES.each do |r|
+            subs << PglogicalSubscription.new(:host => PG_HOST, :port => PG_PORT, :user => PG_USER, :dbname => database(r), :password => PG_PASS)
+          end
+          MiqPglogical.save_global_region(subs, [])
+        end
+
+        private
+
+        def command_environment(region)
+          "REGION='#{region}' RAILS_ENV='development' DATABASE_URL='#{database_url(region)}'"
+        end
+
+        def setup_global_region(region)
+          run_command("#{command_environment(region)} bin/rails r 'TaskHelpers::Development::Replication.setup_global_region_script'")
+          run_command("psql -U #{PG_USER} #{database(region)} -c 'select * from pg_subscription;'")
+        end
+
+        def setup_remote_region(region)
+          run_command("#{command_environment(region)} bin/rails r 'MiqRegion.replication_type= :remote'")
+          run_command("psql -U #{PG_USER} #{database(region)} -c 'select * from pg_publication;'")
+        end
+
+        def setup_one_region(region)
+          run_command("#{command_environment(region)} DISABLE_DATABASE_ENVIRONMENT_CHECK='true' bin/rake evm:db:region")
+          run_command("#{command_environment(region)} bin/rails r 'EvmDatabase.seed_primordial'")
+        ensure
+          FileUtils.rm_f(guid_file)
+        end
+
+        def run_command(command, raise_on_error: true)
+          puts "+" * 50
+          puts "Running: #{command.inspect}..."
+          puts AwesomeSpawn.run!(command).output
```

Comment:
AwesomeSpawn (and Kernel), can take an env as a parameter, so prefer that over building strings for the env.

```ruby
        def command_environment(region)
          {"REGION" => region, "RAILS_ENV" => 'development', "DATABASE_URL" => database_url(region)}
        end
# ...
          run_command("bin/rails r 'TaskHelpers::Development::Replication.setup_global_region_script'", env: command_environment(region))
# ...
        def run_command(command, raise_on_error: true, env:)
          puts AwesomeSpawn.run!(command, env: env).output
```

Something like that.

Additionally, if all you are using AwesomeSpawn for is just to shell out commands (as opposed to preventing CLI injection, hash-like params, etc), then it's just much easier and cleaner to use `system()` directly (which also takes the env)

Reply:
Oh yeah, I forgot about passing the env.

Note, I am using AwesomeSpawn for capturing the exit code and raising by default on non-zero exit codes. and for capturing the output.  Backticks doesn't check exit codes and `system` doesn't return the output. Technically, we don't need the output but I think it's useful to see what's happening.

Reply:
`system` returns a boolean on exit codes, so you can do things like  `raise unless system("blah")`

Reply:
Another reason to use system is that it streams the output, so it doesn't look like it freezes and then suddenly you see everything.

Reply:
Ok, cool, looks better, thanks.  I updated the PR description.